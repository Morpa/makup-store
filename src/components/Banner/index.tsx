import Image from 'next/image'

import Button from 'components/Button'

import * as S from './styles'

export type BannerProps = {
  img: string
  title: string
  buttonLabel: string
  buttonLink: string
}

export const Banner = ({
  img,
  title,
  buttonLabel,
  buttonLink
}: BannerProps) => (
  <S.Wrapper>
    <S.ImageWrapper>
      <Image
        src={img}
        alt={title}
        layout="fill"
        objectFit="cover"
        placeholder="blur"
        blurDataURL={img}
      />
    </S.ImageWrapper>

    <S.Caption>
      <S.Title>{title}</S.Title>
      <Button as="a" href={buttonLink} size="large">
        {buttonLabel}
      </Button>
    </S.Caption>
  </S.Wrapper>
)
