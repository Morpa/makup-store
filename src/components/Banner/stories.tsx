import { Story, Meta } from '@storybook/react/types-6-0'
import { Banner, BannerProps } from '.'

export default {
  title: 'Banner',
  component: Banner,
  args: {
    img: 'https://source.unsplash.com/user/morpa/1040x580',
    title: 'Ultra HD Face Essentials Palette',
    buttonLabel: 'Buy now',
    buttonLink: '/products/essentials-palette'
  },
  parameters: {
    layout: 'fullscreen'
  }
} as Meta

export const Default: Story<BannerProps> = (args) => (
  <div style={{ maxWidth: '104rem', margin: '0 auto' }}>
    <Banner {...args} />
  </div>
)
