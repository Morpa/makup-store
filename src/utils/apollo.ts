import { ApolloClient, HttpLink, NormalizedCacheObject } from '@apollo/client'
import { useMemo } from 'react'

import apolloCache from './apolloCache'

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient() {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql/`
  })

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: httpLink,
    cache: apolloCache
  })
}

export function initializeApollo(initialState = {}) {
  const apolloClientGlobal = apolloClient ?? createApolloClient()

  if (initialState) {
    apolloClientGlobal.cache.restore(initialState)
  }

  if (typeof window === 'undefined') return apolloClientGlobal

  apolloClient = apolloClient ?? apolloClientGlobal

  return apolloClient
}

export function useApollo(initialState = {}) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
