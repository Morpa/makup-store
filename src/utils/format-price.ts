export default function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('pt-PT', {
    style: 'currency',
    currency: 'EUR'
  }).format(price)
}
