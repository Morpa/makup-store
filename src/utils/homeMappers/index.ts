import { ProductApiResponse } from 'services/api'

export const productsMapper = (products: ProductApiResponse[]) => {
  return products
    ? products.map((item) => {
        return {
          id: item.id,
          title: item.name,
          brand: item.brand?.toLocaleUpperCase(),
          img: `https:${item.api_featured_image}`,
          price:
            Number(item.price) > 0
              ? item.price
              : Math.floor(Math.random() * 50),
          rating: Number(item?.rating) > 0 ? Number(item.rating) : Number(0)
        }
      })
    : {}
}

export const bannersMapper = (products: ProductApiResponse[]) => {
  return products
    ? products.map((item) => {
        return {
          img: `https:${item.api_featured_image}`,
          title: item.name,
          buttonLabel: 'Buy now',
          buttonLink: `/product/${item.id}`
        }
      })
    : {}
}

export const highlightMapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  products: any,
  title: string,
  subtitle: string,
  alignment: string
) => {
  return products
    ? products.map((item: ProductApiResponse) => {
        return {
          title,
          subtitle,
          backgroundImage: `https:${item.api_featured_image}`,
          buttonLabel: 'Buy now',
          buttonLink: `/product/${item.id}`,
          alignment
        }
      })
    : {}
}
