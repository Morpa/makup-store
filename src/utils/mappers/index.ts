import { QueryProducts_products_product } from 'graphql/generated/QueryProducts'

export const productsMapper = (products: QueryProducts_products_product[]) => {
  return products
    ? products.map((item) => {
        return {
          id: item.id,
          title: item.name,
          brand: (item.brand?.length && item.brand?.toLocaleUpperCase()) || '',
          img: `https:${item.api_featured_image}`,
          price:
            Number(item.price) > 0
              ? item.price
              : Math.floor(Math.random() * 50),
          rating: Number(item?.rating) > 0 ? Number(item.rating) : Number(0)
        }
      })
    : {}
}

export const bannersMapper = (products: QueryProducts_products_product[]) => {
  return products
    ? products.map((item) => {
        return {
          img: `https:${item.api_featured_image}`,
          title: item.name,
          buttonLabel: 'Buy now',
          buttonLink: `/product/${item.id}`
        }
      })
    : {}
}

export const highlightMapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  products: any,
  title: string,
  subtitle: string,
  alignment: string
) => {
  return products
    ? products.map((item: QueryProducts_products_product) => {
        return {
          title,
          subtitle,
          backgroundImage: `https:${item.api_featured_image}`,
          buttonLabel: 'Buy now',
          buttonLink: `/product/${item.id}`,
          alignment
        }
      })
    : {}
}
