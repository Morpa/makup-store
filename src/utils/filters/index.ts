import { ParsedUrlQueryInput } from 'querystring'

import { ItemProps } from 'components/ExploreSidebar'

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key)
    const isCheckbox = item?.type === 'checkbox'
    const isArray = Array.isArray(queryString[key])

    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key]
  })

  return obj
}
