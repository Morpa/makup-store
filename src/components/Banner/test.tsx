import { render, screen } from 'utils/test-utils'

import Banner from '.'

const props = {
  img: 'https://source.unsplash.com/user/morpa/1042x580',
  title: 'Deadpool',
  subtitle: '<p>Watch the new movie</p>',
  buttonLabel: 'More info',
  buttonLink: '/movies/deadpool'
}

describe('<Banner />', () => {
  it('should render correctly', () => {
    const { container } = render(<Banner {...props} />)

    expect(
      screen.getByRole('heading', { name: /Deadpool/i })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', { name: /Watch the new movie/i })
    ).toBeInTheDocument()

    expect(screen.getByRole('img', { name: /Deadpool/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
