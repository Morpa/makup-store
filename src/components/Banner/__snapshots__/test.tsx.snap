// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c4:focus {
  outline: 1px dashed;
}

.c4:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  position: relative;
}

.c2 {
  width: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 2.4rem;
}

.c3 {
  font-size: 1.8rem;
  font-weight: 600;
  color: #FAFAFA;
}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (min-width:768px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c2 {
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
    position: absolute;
    bottom: 0;
    left: 0;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2.8rem;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Ultra HD Face Essentials Palette"
      layout="fill"
      placeholder="blur"
      src="https://source.unsplash.com/user/morpa/1040x580"
    />
  </div>
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      Ultra HD Face Essentials Palette
    </h2>
    <a
      class="c4"
      href="/products/essentials-palette"
    >
      <span>
        Buy now
      </span>
    </a>
  </div>
</main>
`;
